modules/issue_tracker/ajax/EditIssue.php

```diff
@@ -0,0 +1,634 @@
+        );
+    } else {
+        return array('isValidSubmission' => true);
+    } //phew
+}
+
+/**
+ * Puts updated fields into the issues_history table.
+ *
+ * @param array $issueValues the new values
+ * @param string $issueID the issue ID
+ *
+ * @throws DatabaseException
+ *
+ * @return void
+ */
+function updateHistory($issueValues, $issueID)
+{
+    $user =& User::singleton();
+    $db =& Database::singleton();
+    $undesiredFields = array('lastUpdatedBy');
+
+    error_log($user->getData('UserID'));
+
+    foreach ($issueValues as $key => $value) {
+        if (in_array($key, $undesiredFields)) {
+            continue;
+        }
+        if (!empty($value)) { //check that all kinds of nulls are being dealt with
+            $changedValues = [
+                'newValue' => $value,
+                'fieldChanged' => $key,
+                'issueID' => $issueID,
+                'addedBy' => $user->getData('UserID'),
+            ];
+            $db->insert('issues_history', $changedValues);
+        }
+    }
+}
+
+/**
+ * Will keep track of comment edit history
+ *
+ * @param int $issueCommentID the comment that is being edited
+ * @param string $newCommentValue the new comment
+ *
+ * @throws DatabaseException
+ *
+ * @return void
+ */
+function updateCommentHistory($issueCommentID, $newCommentValue)
+{
+    $user =& User::singleton();
+    $db =& Database::singleton();
+
+    $changedValue = array(
+        'issueCommentID' => $issueCommentID,
+        'newValue' => $newCommentValue,
+        'editedBy' => $user->getData('UserID'),
+    );
+
+    $db->insert('issues_comments_history', $changedValue);
+}
+
+/**
+ * Gets a list of those watching an issue.
+ *
+ * @param int $issueID the relevant issue
+ *
+ * @throws DatabaseException
+ *
+ * @return array those who are watching
+ */
+function getWatching($issueID)
+{
+    $db =& Database::singleton();
+
+    $watching = $db->pselect("SELECT userID from issues_watching WHERE issueID=:issueID",
+        array('issueID' => $issueID));
+
+    $whoIsWatching = array();
+    foreach ($watching as $watcher) {
+        $whoIsWatching[] = $watcher['userID'];
+    }
+    return $whoIsWatching;
+}
+
+/**
+ * Gets the changes to values, and the comments relevant to the given issue
+ *
+ * @param int $issueID the issueID
+ *
+ * @throws DatabaseException
+ *
+ * @return string $commentHistory html
+ */
+function getComments($issueID)
+{
+    $db =& Database::singleton();
+    $unformattedComments = $db->pselect(
+        "SELECT newValue, fieldChanged, dateAdded, addedBy " .
+        "FROM issues_history where issueID=:issueID " .
+        "UNION " .
+        "SELECT issueComment, 'comment', dateAdded, addedBy " .
+        "FROM issues_comments where issueID=:issueID " .
+        "ORDER BY dateAdded",
+        array('issueID' => $issueID)
+    );
+
+    //todo: get real names as well
+
+    $commentHistory = '';
+    foreach ($unformattedComments as $comment) {
```

Comment:
for ($commentRow=0; ++ {
   $tpl_data[$commentHistoryArray][$commentRow][0] = _value_; 
   $tpl_data[$commentHistoryArray][$commentRow][1] = _field_; 
   $tpl_data[$commentHistoryArray][$commentRow][2] = _date_; 
   $tpl_data[$commentHistoryArray][$commentRow][3] = _user_; 
}

and in the template (this not actual smarty, just pseudocode) : 
Foreach row in $commentHistoryArray {
print " $commentRow[3] : $commentRow[2] changed field $commentRow[1]  to <b> $commentRow[0] <b> "; 
}